File: apps/flutter_app/lib/sync/vault_sync_merger.dart
import 'package:password_manager_core/password_manager_core.dart';

import '../state/sync_settings.dart';

enum VersionComparison {
  equal,
  localDominates,
  remoteDominates,
  concurrent,
}

typedef IdGenerator = String Function();

typedef ConflictLabelBuilder = String Function(VaultItem item, bool isRemote);

class MergeStats {
  const MergeStats({
    required this.total,
    required this.conflicts,
    required this.deletes,
  });

  final int total;
  final int conflicts;
  final int deletes;
}

class MergeResult {
  const MergeResult({
    required this.items,
    required this.stats,
  });

  final List<VaultItem> items;
  final MergeStats stats;
}

class VaultSyncMerger {
  VaultSyncMerger({
    required this.idGenerator,
    required this.conflictLabelBuilder,
    required this.conflictStrategy,
  });

  final IdGenerator idGenerator;
  final ConflictLabelBuilder conflictLabelBuilder;
  final ConflictStrategy conflictStrategy;

  MergeResult merge({
    required List<VaultItem> localItems,
    required List<VaultItem> remoteItems,
  }) {
    final merged = <VaultItem>[];
    var conflicts = 0;
    var deletes = 0;

    void addItem(VaultItem item) {
      merged.add(item);
      if (item.isDeleted) {
        deletes += 1;
      }
    }

    final versionCache = <String, Map<String, int>>{};
    Map<String, int> effectiveVersion(VaultItem item) {
      if (item.version.isNotEmpty) {
        return item.version;
      }
      return versionCache.putIfAbsent(item.id, () {
        final updater = item.updatedBy.isEmpty ? 'legacy' : item.updatedBy;
        return {updater: 1};
      });
    }

    final localMap = {
      for (final item in localItems) item.id: item,
    };
    final remoteMap = {
      for (final item in remoteItems) item.id: item,
    };

    for (final entry in localMap.entries) {
      final local = entry.value;
      final remote = remoteMap.remove(entry.key);
      if (remote == null) {
        addItem(local);
        continue;
      }

      final comparison =
          compareVersion(effectiveVersion(local), effectiveVersion(remote));
      if (comparison == VersionComparison.equal) {
        final picked = _pickLatest(local, remote);
        addItem(picked);
        continue;
      }

      if (comparison == VersionComparison.localDominates) {
        addItem(local);
        continue;
      }

      if (comparison == VersionComparison.remoteDominates) {
        addItem(remote);
        continue;
      }

      // concurrent updates
      if (local.isDeleted != remote.isDeleted) {
        conflicts += 1;
        final deleted = local.isDeleted ? local : remote;
        final active = local.isDeleted ? remote : local;
        addItem(deleted);
        merged.add(_conflictClone(active, isRemote: active == remote));
        continue;
      }
      if (_isSamePayload(local, remote)) {
        final picked = _pickLatest(local, remote);
        addItem(picked);
        continue;
      }

      conflicts += 1;
      final primary = _choosePrimary(local, remote);
      final secondary = primary == local ? remote : local;
      merged.add(primary);
      merged.add(_conflictClone(secondary, isRemote: secondary == remote));
    }

    for (final remote in remoteMap.values) {
      addItem(remote);
    }

    return MergeResult(
      items: merged,
      stats: MergeStats(
        total: merged.length,
        conflicts: conflicts,
        deletes: deletes,
      ),
    );
  }

  static VersionComparison compareVersion(
    Map<String, int> local,
    Map<String, int> remote,
  ) {
    var localGreater = false;
    var remoteGreater = false;
    final allKeys = {...local.keys, ...remote.keys};
    for (final key in allKeys) {
      final localValue = local[key] ?? 0;
      final remoteValue = remote[key] ?? 0;
      if (localValue > remoteValue) {
        localGreater = true;
      } else if (remoteValue > localValue) {
        remoteGreater = true;
      }
      if (localGreater && remoteGreater) {
        return VersionComparison.concurrent;
      }
    }
    if (!localGreater && !remoteGreater) {
      return VersionComparison.equal;
    }
    if (localGreater && !remoteGreater) {
      return VersionComparison.localDominates;
    }
    if (!localGreater && remoteGreater) {
      return VersionComparison.remoteDominates;
    }
    return VersionComparison.concurrent;
  }

  VaultItem _pickLatest(VaultItem local, VaultItem remote) {
    if (local.updatedAt.isAtSameMomentAs(remote.updatedAt)) {
      return local;
    }
    return local.updatedAt.isAfter(remote.updatedAt) ? local : remote;
  }

  bool _isSamePayload(VaultItem local, VaultItem remote) {
    if (local.label != remote.label) {
      return false;
    }
    if (local.type != remote.type) {
      return false;
    }
    if (local.isDeleted != remote.isDeleted) {
      return false;
    }
    if (local.kdfIterations != remote.kdfIterations) {
      return false;
    }
    if (local.metadataCategory != remote.metadataCategory) {
      return false;
    }
    if (!_sameStringList(local.metadataTags, remote.metadataTags)) {
      return false;
    }
    if (local.encryptedPayload.version != remote.encryptedPayload.version) {
      return false;
    }
    return _bytesEqual(local.encryptedPayload.ciphertext,
            remote.encryptedPayload.ciphertext) &&
        _bytesEqual(
            local.encryptedPayload.nonce, remote.encryptedPayload.nonce) &&
        _bytesEqual(local.encryptedPayload.mac, remote.encryptedPayload.mac);
  }

  bool _sameStringList(List<String> a, List<String> b) {
    if (a.length != b.length) {
      return false;
    }
    for (var i = 0; i < a.length; i++) {
      if (a[i] != b[i]) {
        return false;
      }
    }
    return true;
  }

  bool _bytesEqual(List<int> a, List<int> b) {
    if (a.length != b.length) {
      return false;
    }
    var diff = 0;
    for (var i = 0; i < a.length; i++) {
      diff |= a[i] ^ b[i];
    }
    return diff == 0;
  }

  VaultItem _choosePrimary(VaultItem local, VaultItem remote) {
    switch (conflictStrategy) {
      case ConflictStrategy.localWins:
        return local;
      case ConflictStrategy.remoteWins:
        return remote;
      case ConflictStrategy.keepBoth:
        return _pickLatest(local, remote);
    }
  }

  VaultItem _conflictClone(VaultItem source, {required bool isRemote}) {
    final labelSuffix = conflictLabelBuilder(source, isRemote);
    final now = DateTime.now().toUtc();
    final updatedBy = source.updatedBy.isEmpty ? 'legacy' : source.updatedBy;
    final baseVersion =
        source.version.isNotEmpty ? source.version[updatedBy] ?? 1 : 1;
    return VaultItem(
      id: idGenerator(),
      label: '${source.label} $labelSuffix',
      type: source.type,
      encryptedPayload: source.encryptedPayload,
      kdfSalt: source.kdfSalt,
      kdfIterations: source.kdfIterations,
      createdAt: now,
      updatedAt: source.updatedAt,
      version: {updatedBy: baseVersion},
      updatedBy: updatedBy,
      isDeleted: source.isDeleted,
      deletedAt: source.deletedAt,
      metadataCategory: source.metadataCategory,
      metadataTags: source.metadataTags,
    );
  }
}

File: apps/android_native/app/src/main/java/com/example/passwordmanagernative/sync/VaultSyncMerger.kt
package com.example.passwordmanagernative.sync

import com.example.passwordmanagernative.model.VaultEntry
import java.time.Instant
import java.util.UUID

enum class VersionComparison {
    EQUAL,
    LOCAL_DOMINATES,
    REMOTE_DOMINATES,
    CONCURRENT,
}

enum class SyncConflictStrategy {
    LOCAL_WINS,
    REMOTE_WINS,
    KEEP_BOTH,
}

data class SyncMergeStats(
    val total: Int,
    val conflicts: Int,
    val deletes: Int,
)

data class SyncMergeResult(
    val entries: List<VaultEntry>,
    val stats: SyncMergeStats,
)

class VaultSyncMerger(
    private val idGenerator: () -> String = { UUID.randomUUID().toString() },
    private val conflictLabelBuilder: (VaultEntry, Boolean) -> String = { _, _ -> "(conflict)" },
    private val conflictStrategy: SyncConflictStrategy = SyncConflictStrategy.KEEP_BOTH,
    private val clock: () -> Instant = { Instant.now() },
) {
    fun merge(localEntries: List<VaultEntry>, remoteEntries: List<VaultEntry>): SyncMergeResult {
        val merged = mutableListOf<VaultEntry>()
        var conflicts = 0
        var deletes = 0

        fun addEntry(entry: VaultEntry) {
            merged += entry
            if (entry.isDeleted) {
                deletes += 1
            }
        }

        val versionCache = mutableMapOf<String, Map<String, Int>>()
        fun effectiveVersion(entry: VaultEntry): Map<String, Int> {
            if (entry.version.isNotEmpty()) return entry.version
            return versionCache.getOrPut(entry.id) {
                mapOf((entry.updatedBy.ifBlank { LEGACY_UPDATER }) to 1)
            }
        }

        val remoteById = remoteEntries.associateBy { it.id }.toMutableMap()
        localEntries.associateBy { it.id }.forEach { (id, local) ->
            val remote = remoteById.remove(id)
            if (remote == null) {
                addEntry(local)
                return@forEach
            }

            when (compareVersion(effectiveVersion(local), effectiveVersion(remote))) {
                VersionComparison.EQUAL -> addEntry(pickLatest(local, remote))
                VersionComparison.LOCAL_DOMINATES -> addEntry(local)
                VersionComparison.REMOTE_DOMINATES -> addEntry(remote)
                VersionComparison.CONCURRENT -> {
                    if (local.isDeleted != remote.isDeleted) {
                        conflicts += 1
                        val deleted = if (local.isDeleted) local else remote
                        val active = if (local.isDeleted) remote else local
                        addEntry(deleted)
                        merged += conflictClone(active, isRemote = active === remote)
                    } else if (samePayload(local, remote)) {
                        addEntry(pickLatest(local, remote))
                    } else {
                        conflicts += 1
                        val primary = choosePrimary(local, remote)
                        val secondary = if (primary === local) remote else local
                        merged += primary
                        merged += conflictClone(secondary, isRemote = secondary === remote)
                    }
                }
            }
        }

        remoteById.values.forEach(::addEntry)

        return SyncMergeResult(
            entries = merged,
            stats = SyncMergeStats(
                total = merged.size,
                conflicts = conflicts,
                deletes = deletes,
            ),
        )
    }

    private fun pickLatest(local: VaultEntry, remote: VaultEntry): VaultEntry =
        if (local.updatedAt >= remote.updatedAt) local else remote

    private fun samePayload(local: VaultEntry, remote: VaultEntry): Boolean =
        local.label == remote.label &&
            local.type == remote.type &&
            local.payload == remote.payload &&
            local.isDeleted == remote.isDeleted

    private fun choosePrimary(local: VaultEntry, remote: VaultEntry): VaultEntry =
        when (conflictStrategy) {
            SyncConflictStrategy.LOCAL_WINS -> local
            SyncConflictStrategy.REMOTE_WINS -> remote
            SyncConflictStrategy.KEEP_BOTH -> pickLatest(local, remote)
        }

    private fun conflictClone(source: VaultEntry, isRemote: Boolean): VaultEntry {
        val updatedBy = source.updatedBy.ifBlank { LEGACY_UPDATER }
        val baseVersion = if (source.version.isNotEmpty()) {
            source.version[updatedBy] ?: 1
        } else {
            1
        }
        return source.copy(
            id = idGenerator(),
            label = "${source.label} ${conflictLabelBuilder(source, isRemote)}",
            createdAt = clock(),
            version = mapOf(updatedBy to baseVersion),
            updatedBy = updatedBy,
        )
    }

    companion object {
        private const val LEGACY_UPDATER = "legacy"

        fun compareVersion(local: Map<String, Int>, remote: Map<String, Int>): VersionComparison {
            var localGreater = false
            var remoteGreater = false
            (local.keys + remote.keys).forEach { key ->
                val localValue = local[key] ?: 0
                val remoteValue = remote[key] ?: 0
                if (localValue > remoteValue) {
                    localGreater = true
                } else if (remoteValue > localValue) {
                    remoteGreater = true
                }
                if (localGreater && remoteGreater) {
                    return VersionComparison.CONCURRENT
                }
            }
            return when {
                !localGreater && !remoteGreater -> VersionComparison.EQUAL
                localGreater -> VersionComparison.LOCAL_DOMINATES
                remoteGreater -> VersionComparison.REMOTE_DOMINATES
                else -> VersionComparison.CONCURRENT
            }
        }
    }
}
